sandbox/grist/imports/import_csv.py

```diff
@@ -90,108 +90,114 @@ def parse_file(file_path, parse_options=None):
   """
   parse_options = parse_options or {}
 
-  with open(file_path, "rb") as f:
+  with codecs.open(file_path, "rb") as f:
+    sample = f.read(100000)
+  encoding = chardet.detect(sample)['encoding'] or "utf8"
+  # In addition, always prefer UTF8 over ASCII.
+  if encoding == 'ascii':
+    encoding = 'utf8'
+  log.info("Using encoding %s" % encoding)
+
+  with codecs.open(file_path, mode="r", encoding=encoding) as f:
     parsing_options, export_list = _parse_open_file(f, parse_options=parse_options)
     return parsing_options, export_list
 
 
+def _guess_dialect(file_obj):
+  try:
+    # Restrict allowed delimiters to prevent guessing other char than this list.
+    dialect = csv.Sniffer().sniff(file_obj.read(100000), delimiters=['\t', ',', ';', '|'])
+    log.info("Guessed dialect %s" % dict(dialect.__dict__))
+    return dialect
+  except csv.Error:
+    log.info("Cannot guess dialect using Excel as fallback.")
+    return csv.excel
+  finally:
+    file_obj.seek(0)
+
 def _parse_open_file(file_obj, parse_options=None):
-  options = {}
   csv_keys = ['delimiter', 'quotechar', 'lineterminator', 'doublequote', 'skipinitialspace']
-  csv_options = {k: parse_options.get(k) for k in csv_keys}
-  if six.PY2:
-    csv_options = {k: v.encode('utf8') if isinstance(v, six.text_type) else v
-                   for k, v in csv_options.items()}
-
-  table_set = messytables.CSVTableSet(file_obj,
-                                      delimiter=csv_options['delimiter'],
-                                      quotechar=csv_options['quotechar'],
-                                      lineterminator=csv_options['lineterminator'],
-                                      doublequote=csv_options['doublequote'],
-                                      skipinitialspace=csv_options['skipinitialspace'])
+  options = {}
+  dialect = _guess_dialect(file_obj)
 
-  num_rows = parse_options.get('NUM_ROWS', 0)
+  csv_options = {}
+  for key in csv_keys:
+    value = parse_options.get(key) or getattr(dialect, key, None)
```

Comment:
```suggestion
    value = parse_options.get(key, getattr(dialect, key, None))
```

That way `doublequote=False` in the options will take precedence over the guessed dialect.

Reply:
Good catch, thanks!